### DISCUSSION TITLE
how to package the app to others

### DISCUSSION THREAD

**Discussion by Jzhenli** [termux/termux-app#3276] [Q&A] [answered]
Hi, I'm beginner to this project, now I have a question need your help.

I have installed termux on my phone, write a python flask server on it, and it works well. Now I want to let this app can run on another phone, how to do that? 
I know i can setup it again on the second phone as the first one, but it seems too complicated. If there is an easy way to do this?  Such as to package all the things to one apk, or something like this.

Thanks.

**Comment by agnostic-apollo** [comment, marked as answer]
If both phones have same architecture as per `dpkg --print-architecture`, then do backup and restore.

https://wiki.termux.com/wiki/Backing_up_Termux

**Reply by Jzhenli** [reply]
thanks for your reply, it's helpful.

**Reply by agnostic-apollo** [reply]
Welcome.

**Reply by Jzhenli** [reply]
Hi, if some get the phone, then they can easily get all the files on the termux? I'm not sure how to keep them safety.

**Reply by agnostic-apollo** [reply]
Put screen lock so that data partition is encrypted. Or for additional encryption, use encrypted images with root.

https://glow.li/posts/encrypt-termux-fs

**Reply by Jzhenli** [reply]
Hi, for the unrooted device, if it's there's a way for user to input username and password when open the seesion in termux app?

**Reply by sylirre** [reply]
@Jzhenli No, that's not possible for user name. Android automatically generates the Linux user name from user ID on-fly whenever program calls respective functions such as `getpwuid()`. Each application has its own Linux user id, e.g. UID `10352` which would be converted to user name `u0_a352`. You cannot change this behavior of Android OS as it is coded inside its core component (libc).

As for password, things are a bit easier. You can implement "password login" through adding some code inside `/data/data/com.termux/files/usr/etc/termux-login.sh`. For example (bare minimum):
```
read -res -p "Enter your password: " PASSWORD
if [ "$PASSWORD" != "mypassword" ]; then
    echo "Invalid password."
    exit 1
fi
```

However know that all kinds of local password login are not supported by design. The device screen lock is sufficient. Alternate method is to setup application locking through third party apps. Other methods **will not make your Termux** secure as the latter is designed in way that user can modify it and take the data out even if the app is completely broken or locked down. The failsafe shell will bypass all kinds of password authentication: https://wiki.termux.com/wiki/Recover_a_broken_environment. How to access Termux files without root and various passwords is explained here: https://wiki.termux.com/wiki/Internal_and_external_storage#Access_Termux_from_a_file_manager

Therefore I do not recommend trying to find a way to configure a "password login". It just will give you a false sense of security.

On other hand it is possible to enable password authentication for SSH: https://wiki.termux.com/wiki/Remote_Access#Setting_up_password_authentication
